### PR TITLE
Avoid concurrent use of ConditionalWeakTable on net472

### PR DESCRIPTION
**Pull request by sharwell**
Avoids a read lock in ConditionalWeakTable for .NET Framework, which results in extreme contention on concurrent paths. For source generators producing large amounts of code in parallel, the default use of ElasticMarker made most SyntaxFactory methods completely unusable (e.g. CsWin32 wrote a complete new syntax factory just to avoid this).

The clear downside of this change is it increases the size of `GreenNode` by 8 bytes when executing on .NET Framework (including inside Visual Studio), which is _heavily_ allocated.

Fixes [AB#1420484](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1420484)

### CODE REVIEW

**Comment by sharwell** [comment]
This is going to be a tough call since the change offers significant benefits to one scenario with a non-trivial disadvantage across all .NET Framework usage. The problems resolve themselves when .NET Core / .NET is used, but some environments including Visual Studio and certain builds still use .NET Framework.

/cc @jaredpar @cston @VSadov 

/cc @davkean @AArnott

**Comment by jaredpar** [comment]
Seems like we should start a permanent exception thread with the VS team before taking this PR. Seemingly we need to decide between the idea of taking a higher allocation hit here or living with the contention issue on .NET Framework. Those are the only viable options that we have on the table right now. 

Certainly open to other ideas here but thus far those seem like the only two viable ones.

**Comment by davkean** [comment]
This has a pretty wild stats in the wild, it's ranked #3 of all thread pool issues in 16.11:

Failure | Score | Session % | Cab Count | Threads 90th | Threads 50th | Avg Threads
-- | -- | -- | -- | -- | -- | --
Stk_ThdPl_Microsoft_CodeAnalysis!Microsoft.CodeAnalysis.GreenNode.GetAnnotations() | 313 | 0.06 | 16 | 260| 11 | 66

While it's not seen in a large number of cabs, when it hits it hits hard.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 1]
```diff
@@ -595,8 +610,11 @@ public SyntaxAnnotation[] GetAnnotations()
         {
             if (this.ContainsAnnotations)
             {
-                SyntaxAnnotation[]? annotations;
-                if (s_annotationsTable.TryGetValue(this, out annotations))
+#if NETCOREAPP
+                if (s_annotationsTable.TryGetValue(this, out var annotations))
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:614]: >var [](http://example.com/codeflow?start=61&length=3)

Please spell out the type.

Comment (sharwell) [reply src/Compilers/Core/Portable/Syntax/GreenNode.cs]: ➡️ Changed. If this is a common item for code reviews in this path, to save time and make reviews less negative consider updating **.editorconfig** settings to reveal the situation during initial development.

**Pull request closed without merging**